pkg/envtest/crd.go

```diff
@@ -0,0 +1,200 @@
+	}
+}
+
+// WaitForCRDs waits for the CRDs to appear in discovery
+func WaitForCRDs(config *rest.Config, crds []*apiextensionsv1beta1.CustomResourceDefinition, options CRDOptions) error {
+	// Add each CRD to a map of GroupVersion to Resource
+	waitingFor := map[schema.GroupVersion]*sets.String{}
+	for _, crd := range crds {
+		gv := schema.GroupVersion{Group: crd.Spec.Group, Version: crd.Spec.Version}
+		if _, found := waitingFor[gv]; !found {
+			// Initialize the set
+			waitingFor[gv] = &sets.String{}
+		}
+		// Add the Resource
+		waitingFor[gv].Insert(crd.Spec.Names.Plural)
+	}
+	// Poll until all resources are found
+	return wait.PollImmediate(options.pollInterval, options.maxTime, func() (bool, error) {
+		// Create a new clientset to avoid any client caching of discovery
+		cs, err := clientset.NewForConfig(config)
+		if err != nil {
+			return false, err
+		}
+
+		allFound := true
+		for gv, resources := range waitingFor {
+			// All resources found, do nothing
+			if resources.Len() == 0 {
+				delete(waitingFor, gv)
+				continue
+			}
+
+			// Get the Resources for this GroupVersion
+			resourceList, err := cs.Discovery().ServerResourcesForGroupVersion(gv.Group + "/" + gv.Version)
+			if err != nil {
+				return false, nil
+			}
+
+			// Remove each found resource from the resources set that we are waiting for
+			for _, resource := range resourceList.APIResources {
+				resources.Delete(resource.Name)
+			}
+
+			// Still waiting on some resources in this group version
+			if resources.Len() != 0 {
+				allFound = false
+			}
+		}
+		return allFound, nil
+	})
+}
+
+// CreateCRDs creates the CRDs
+func CreateCRDs(config *rest.Config, crds []*apiextensionsv1beta1.CustomResourceDefinition) error {
+	cs, err := clientset.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	// Create each CRD
+	for _, crd := range crds {
+		if _, err := cs.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// readCRDs reads the CRDs from files and Unmarshals them into structs
+func readCRDs(path string) ([]*apiextensionsv1beta1.CustomResourceDefinition, error) {
+	// Get the CRD files
+	var files []os.FileInfo
+	var err error
+	if files, err = ioutil.ReadDir(path); err != nil {
+		return nil, err
+	}
+
+	// White list the file extensions that may contain CRDs
+	crdExts := sets.NewString(".json", ".yaml")
+
+	var crds []*apiextensionsv1beta1.CustomResourceDefinition
+	for _, file := range files {
+		// Only parse whitelisted file types
+		if !crdExts.Has(filepath.Ext(file.Name())) {
+			continue
+		}
+
+		// Unmarshal the file into a struct
+		b, err := ioutil.ReadFile(filepath.Join(path, file.Name()))
+		if err != nil {
+			return nil, err
+		}
+		crd := &apiextensionsv1beta1.CustomResourceDefinition{}
+		if err = yaml.Unmarshal(b, crd); err != nil {
+			return nil, err
+		}
+
+		// Check that it is actually a CRD
+		if crd.Spec.Names.Kind == "" || crd.Spec.Group == "" {
```

Comment:
Should take`Spec.Scope` to consideration as well ?

Reply:
I don't thin so.  This is really just to check if it was an unrelated type that happened to parse since go doesn't throw an error if the file is valid yaml but the schema doesn't match.